src/nupic/utils/SlidingWindow.hpp

```diff
@@ -0,0 +1,163 @@
+/* Numenta Platform for Intelligent Computing (NuPIC)
+ * Copyright (C) 2016, Numenta, Inc.  Unless you have an agreement
+ * with Numenta, Inc., for a separate license for this software code, the
+ * following terms and conditions apply:
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ *
+ * http://numenta.org/licenses/
+ * ---------------------------------------------------------------------
+ */
+
+
+#ifndef NUPIC_UTIL_SLIDING_WINDOW_HPP
+#define NUPIC_UTIL_SLIDING_WINDOW_HPP
+
+#include <vector>
+
+#include <nupic/types/Types.hpp>
+#include <nupic/utils/Log.hpp>
+
+namespace nupic {
+  namespace util {
+
+    template<class T> 
+    class SlidingWindow {
+      public:
+        SlidingWindow(UInt maxCapacity);
+        SlidingWindow(UInt maxCapacity, std::vector<T> initialData);
```

Comment:
In my code above, "Iterator" is just the "T" of the template. I could have called it "InputIterator42" or "George".

To write this method, you don't have to know anything about iterators. Just write it as if they're pointers.

~~~c++
for(Iterator it = initialData_begin; it != initialData_end; ++it)
{
  append(*it);
}
~~~

Separately, you could choose to make the MovingWindow iterable, but that's a separate feature. You could create a nested class `MovingWindow::Iterator` and add `begin` and `end` methods to the `MovingWindow`, but that's optional. To do this, you don't need to know anything about `std::iterator`, you just need to implement a few public methods. You can see an example where I did this here: https://github.com/numenta/nupic.core/blob/b83798346e85bdbf773652c7597f00a4a9ed6575/src/nupic/utils/GroupBy.hpp#L93 (though these classes are kinda complicated)